File: 황재연 (Python)/6주차/[프로그래머스] 미로 탈출 명령어.py
def solution(n, m, x, y, r, c, k):
    
    answer = ''
    
    if (abs(x-r) + abs(y-c)) > k or (abs(x-r) + abs(y-c)) % 2 != k % 2:
        return "impossible"
    
    alpa = ['d', 'l', 'r', 'u']
    
    dx = [1, 0, 0, -1]
    dy = [0, -1, 1, 0]
    

    posX = x
    posY = y

    move = 0

    while (True):
        if posX == r and posY == c and move == k:
            break
        for i in range(4):
            if 0 < posX + dx[i] <= n and 0 < posY + dy[i] <= m:
                if abs(posX + dx[i] - r) + abs(posY + dy[i] - c) <= k - move:
                    print(alpa[i])
                    answer += alpa[i]
                    posX += dx[i]
                    posY += dy[i]
                    move += 1
                    break

    return answer

## d 아래 ㅣ 왼  r 오른 u 위

n, m, x, y, r, c, k = [3,3,1,2,3,3,4]
n, m, x, y, r, c, k = [3,4,2,3,3,1,5]
#n, m, x, y, r, c, k = [6, 6, 2, 6, 6, 5, 11]
#n, m, x, y, r, c, k = [3, 3, 1, 1, 3, 3, 8]


print(solution(n, m, x, y, r, c, k))

File: 황용진 (Javascript)/6주차/[프로그래머스]미로탈출명령어.js
const move = [
  [1, 0],
  [0, -1],
  [0, 1],
  [-1, 0],
]; // 하 좌 우 상
const moveDir = ["d", "l", "r", "u"];

let route = "";

function dfs(n, m, x, y, r, c, cnt, curRoute, k) {
  if (route) return;
  if (cnt + Math.abs(x - r) + Math.abs(y - c) > k) return;
  if (cnt === k) {
    if (x === r && y === c) {
      route = curRoute;
    }
    return;
  }

  for (let i = 0; i < 4; i++) {
    const [dx, dy] = [x + move[i][0], y + move[i][1]];

    if (dx > 0 && dx <= n && dy > 0 && dy <= m) {
      dfs(n, m, dx, dy, r, c, cnt + 1, curRoute + moveDir[i], k);
    }
  }
}

function solution(n, m, x, y, r, c, k) {
  var answer = "";

  let fastAnswer = k - (Math.abs(x - r) + Math.abs(y - c));

  if (fastAnswer < 0 || fastAnswer % 2 !== 0) {
    answer = "impossible";
    return answer;
  }

  dfs(n, m, x, y, r, c, 0, "", k);

  if (!route) {
    answer = "impossible";
    return answer;
  }

  answer = route;

  return answer;
}
